driver-reactive-streams/src/main/com/mongodb/reactivestreams/client/internal/BatchCursorFlux.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.reactivestreams.client.internal;
+
+import com.mongodb.lang.Nullable;
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscriber;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.FluxSink;
+import reactor.core.publisher.Mono;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+
+class BatchCursorFlux<T> implements Publisher<T> {
+
+    private final BatchCursorPublisher<T> batchCursorPublisher;
+    private final Integer originalBatchSize;
+    private final Runnable resetBatchSize;
+    private final AtomicBoolean inProgress = new AtomicBoolean(false);
+    private final AtomicLong demandDelta = new AtomicLong(0);
+    private BatchCursor<T> batchCursor;
+    private FluxSink<T> sink;
+
+    BatchCursorFlux(final BatchCursorPublisher<T> batchCursorPublisher,
+                    @Nullable final Integer originalBatchSize,
+                    final Runnable resetBatchSize) {
+        this.batchCursorPublisher = batchCursorPublisher;
+        this.originalBatchSize = originalBatchSize;
+        this.resetBatchSize = resetBatchSize;
+    }
+
+    @Override
+    public void subscribe(final Subscriber<? super T> subscriber) {
+        Flux.<T>create(sink -> {
+            this.sink = sink;
+            sink.onRequest(l -> {
+                if (demandDelta.addAndGet(l) > 0 && inProgress.compareAndSet(false, true)) {
```

Comment:
As this is backed by a the OverflowStrategy.BUFFER, only call getMore if needed:  `(demandDelta.addAndGet(l) > 0`